src/v/lsm/lsm.h

```diff
@@ -151,9 +163,32 @@ class write_batch {
     // REQUIRES: offsets must be monotonically increasing as added to the batch.
     void remove(std::string_view key, model::offset);
 
+    // Lookup a value in the database as if the current write batch was applied.
+    //
+    // The returned future must finish resolving before the write_batch is
+    // applied to the database.
+    ss::future<std::optional<iobuf>> get(std::string_view key);
+
+    // Create an iterator over the database as if the current write batch was
+    // applied.
+    //
+    // This iterator gives snapshot isolation, so  writes applied after
+    // this future completes will not be seen by the iterator as long as
+    // these writes do not have the same offset.
+    //
+    // The resulting iterator is safe to use concurrently with additional writes
+    // being applied to the batch.
+    //
+    // The returned iterator must not be used after the write_batch is applied
+    // to the database.
```

Comment:
We can remove this limitation if we want, but it will be more expensive, because it will require sharing/copying the data from the memtable to the second.